pkg/orchestrator/controller/controller.go

```diff
@@ -0,0 +1,264 @@
+// hold the ID of the cluster from which the resource originated, allowing us to only process watch notifications
+// emanating from the originating cluster.
+const clusterIDLabelKey = "submariner-io/clusterID"
+
+var (
+	submarinerNamespace string
+)
+
+func init() {
+	flag.StringVar(&submarinerNamespace, "submariner-namespace", "submariner", "The namespace in which the submariner components are deployed.")
+}
+
+func New(federator federate.Federator) *controller {
+	return &controller{
+		federator:       federator,
+		clusterWatchers: make(clusterWatcherMap),
+
+		newSubmClientset: func(c *rest.Config) (submarinerClientset.Interface, error) {
+			return submarinerClientset.NewForConfig(c)
+		},
+	}
+}
+
+func (c *controller) Start() error {
+	err := c.federator.WatchClusters(c)
+	if err != nil {
+		return fmt.Errorf("Could not register cluster watch: %v", err)
+	}
+
+	return nil
+}
+
+func (c *controller) Stop() {
+	c.Lock()
+	defer c.Unlock()
+
+	for _, clusterWatcher := range c.clusterWatchers {
+		clusterWatcher.close()
+	}
+}
+
+func (c *controller) removeExistingClusterWatcher(clusterID string) {
+	watcher := c.clusterWatchers[clusterID]
+	if watcher != nil {
+		klog.V(2).Infof("A watcher for cluster %s already exists - stopping watch", clusterID)
+		watcher.close()
+		delete(c.clusterWatchers, clusterID)
+	}
+}
+
+func (c *controller) startNewWatcherForCluster(kubeConfig *rest.Config, clusterID string) {
+	submarinerClient, err := c.newSubmClientset(kubeConfig)
+	if err != nil {
+		klog.Errorf("error building submariner clientset for cluster %s: %v", clusterID, err)
+		return
+	}
+
+	c.Lock()
+	defer c.Unlock()
+
+	c.removeExistingClusterWatcher(clusterID)
+
+	watcher := &clusterWatcher{
+		clusterID: clusterID,
+		stopChan:  make(chan struct{}),
+		clusterWorkQueue: workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(),
+			fmt.Sprintf("Cluster watcher for %s", clusterID)),
+		endpointWorkQueue: workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(),
+			fmt.Sprintf("Endpoint watcher for %s", clusterID)),
+	}
+	c.clusterWatchers[clusterID] = watcher
+
+	submarinerInformerFactory := submarinerInformers.NewSharedInformerFactoryWithOptions(submarinerClient, 0,
+		submarinerInformers.WithNamespace(submarinerNamespace))
+
+	submV1 := submarinerInformerFactory.Submariner().V1()
+	clusterInformer := submV1.Clusters().Informer()
+	clusterInformer.AddEventHandler(newResourceEventHandler(clusterID, watcher.clusterWorkQueue))
+
+	endpointInformer := submV1.Endpoints().Informer()
+	endpointInformer.AddEventHandler(newResourceEventHandler(clusterID, watcher.endpointWorkQueue))
+
+	submarinerInformerFactory.Start(watcher.stopChan)
+
+	go runQueueWorker(clusterID, c.federator, clusterInformer, watcher.clusterWorkQueue)
+	go runQueueWorker(clusterID, c.federator, endpointInformer, watcher.endpointWorkQueue)
+}
+
+func getLabel(obj interface{}, key string) (string, bool) {
```

Comment:
nit: currently it's only used with a constant so key can be removed as parameter